README.md

```diff
@@ -1,2 +1,393 @@
+    {
+    }
+}
+```
+
+Ballerina -
+```ballerina
+public type DataStore object {
+    isolated function add(string userId, string productId, int quantity);
+
+    isolated function emptyCart(string userId);
+
+    isolated function getCart(string userId) returns Cart;
+};
+
+public isolated class InMemoryStore {
+    *DataStore;
+    private map<Cart> store = {};
+
+    isolated function add(string userId, string productId, int quantity) {
+    }
+}
+
+public isolated class RedisStore {
+    *DataStore;
+
+    isolated function add(string userId, string productId, int quantity) {
+    }
+}
+```
+
+## Search products using query expressions - Catalog Service
+Product catalog service contains all the details of avaiable products. The requirement is to get the products similar to the search query. You can find the original implementation below.
+
+```go
+func (p *productCatalog) SearchProducts(ctx context.Context, req *pb.SearchProductsRequest) (*pb.SearchProductsResponse, error) {
+	time.Sleep(extraLatency)
+	// Intepret query as a substring match in name or description.
+	var ps []*pb.Product
+	for _, p := range parseCatalog() {
+		if strings.Contains(strings.ToLower(p.Name), strings.ToLower(req.Query)) ||
+			strings.Contains(strings.ToLower(p.Description), strings.ToLower(req.Query)) {
+			ps = append(ps, p)
+		}
+	}
+	return &pb.SearchProductsResponse{Results: ps}, nil
+}
+```
+
+Even tough we can implement the same using ballerina using foreach statement, we have used ballerina query expression to implement the search function. Query expressions contain a set of clauses similar to SQL to process the data.
+```ballerina
+remote function SearchProducts(SearchProductsRequest value) returns SearchProductsResponse|error {
+    return {
+        results: from Product product in self.products
+            where isProductRelated(product, value.query)
+            select product
+    };
+}
+
+isolated function isProductRelated(Product product, string query) returns boolean {
+    string queryLowercase = query.toLowerAscii();
+    return product.name.toLowerAscii().includes(queryLowercase) || product.description.toLowerAscii().includes(queryLowercase);
+}
+```
+
+You can read more about Query expressions in this blog. You can have much more complicated queries using limit,let keywords, ordering, joins and so on. You can use query expressions not only for arrays but for streams, and tables as well.
+
+
+## Concurrency safety - Ad Service
+Ballerina is designed for network based applications. The concept of isolation in Ballerina simplifies development by ensuring the safety of shared resources during concurrent execution. Ballerina Compiler warns if the application is not concurrent safe and helps to make it concurrent safe and performant at the same time. The following code shows how a class is marked as readonly so by default compiler makes enables concurrent calls to its objects.
```

Comment:
```suggestion
Ballerina is designed for network-based applications. The concept of isolation in Ballerina simplifies development by ensuring the safety of shared resources during concurrent execution. Ballerina Compiler warns if the application is not concurrent safe and helps to make it concurrent safe and performant at the same time. The following code shows how a class is marked as readonly so that by default, the compiler makes concurrent calls to its objects.
```